validator-api/src/coconut/deposit.rs

```diff
@@ -0,0 +1,430 @@
+// Copyright 2022 - Nym Technologies SA <contact@nymtech.net>
+// SPDX-License-Identifier: Apache-2.0
+
+use coconut_bandwidth_contract::events::{
+    DEPOSITED_FUNDS_EVENT_TYPE, DEPOSIT_ENCRYPTION_KEY, DEPOSIT_INFO, DEPOSIT_VALUE,
+    DEPOSIT_VERIFICATION_KEY,
+};
+use coconut_interface::BlindSignRequestBody;
+use credentials::coconut::bandwidth::BandwidthVoucher;
+use crypto::asymmetric::encryption;
+use crypto::asymmetric::identity::{self, Signature};
+use validator_client::nymd::TxResponse;
+
+use super::error::{CoconutError, Result};
+
+pub async fn extract_encryption_key(
+    blind_sign_request_body: &BlindSignRequestBody,
+    tx: TxResponse,
+) -> Result<encryption::PublicKey> {
+    let blind_sign_request = blind_sign_request_body.blind_sign_request();
+    let public_attributes = blind_sign_request_body.public_attributes();
+    let public_attributes_plain = blind_sign_request_body.public_attributes_plain();
+
+    if !BandwidthVoucher::verify_against_plain(&public_attributes, public_attributes_plain) {
+        return Err(CoconutError::InconsistentPublicAttributes);
+    }
+
+    let tx_hash_str = blind_sign_request_body.tx_hash();
+    let mut message = blind_sign_request.to_bytes();
+    message.extend_from_slice(tx_hash_str.as_bytes());
+
+    let signature = Signature::from_base58_string(blind_sign_request_body.signature())?;
+
+    let attributes: &Vec<_> = tx
```

Comment:
I'll respond to this and the "main" comment here, as it's better threaded. The reasons for using an event approach instead of a storage approach are: at some point, there was an idea of watching the chain for events (which we've abandoned because of the websocket client complexity), and because it _should_ be more lightweight to keep those things forever in a low-level event, then in the more flexible, high-level cosmwasm storage. The query for data thing will be part of the double-spending storage anyway, but that at least needs a faster, overall view of the state at a certain point, whereas here, pin-pointing the query to a block is more then enough for our needs.